core/src/main/java/org/mule/runtime/core/internal/message/DefaultMessageBuilder.java

```diff
@@ -157,170 +155,170 @@ public DefaultMessageBuilder(org.mule.runtime.api.message.Message message) {
 
   @Override
   public CollectionBuilder itemMediaType(MediaType mediaType) {
-    if (dataType instanceof DefaultCollectionDataType) {
-      dataType =
-          ((DataTypeBuilder.DataTypeCollectionTypeBuilder) DataType.builder(this.dataType)).itemMediaType(mediaType).build();
+    if (payload.getDataType() instanceof DefaultCollectionDataType) {
+      payload = new TypedValue(payload.getValue(),
+                               ((DataTypeBuilder.DataTypeCollectionTypeBuilder) builder(payload.getDataType()))
+                                   .itemMediaType(mediaType).build());
     } else {
       throw new IllegalStateException("Item MediaType cannot be set, because payload is not a collection");
     }
     return this;
   }
 
   @Override
-  public InternalMessage.Builder attributes(TypedValue<?> typedValue) {
-    this.attributes = typedValue.getValue();
-    this.attributesDataType = typedValue.getDataType();
+  public InternalMessage.CollectionBuilder attributes(TypedValue<?> attributes) {
```

Comment:
This `CollectionBuilder` is only returned when `collection` methods are used in the interface. Here, in the builder implementation all methods return the most generic builder simply the implementation given the user is using it via interface.

It's not more complex than before, but rather following the same approach with bug fixed. In fact i remove the AttributesBuilder which wasn't being used.